### PR TITLE
Ft questions resources

### PR DESCRIPTION
**Pull request by andela-aadepoju**

### CODE REVIEW

File: app/models/question.rb [hunk 1]
```diff
@@ -12,6 +12,10 @@ def with_answers
       includes(:answers)
     end
 
+    def top
+      where("votes > ?", 0).order("votes DESC")
```
Comment (andela-ooranagwa) [inline app/models/question.rb]: Hmmm ... 
I don't think there is any need for the `where` condition we are more interested in the `ordering` of the votes e.g

> order("votes DESC")

File: app/controllers/questions_controller.rb [hunk 2]
```diff
@@ -0,0 +1,49 @@
+class QuestionsController < ApplicationController
+  before_action :set_question, only: [:show, :update, :destroy]
+
+  def index
+    questions = Question.all
+    render json: questions
```
Comment (andela-ooranagwa) [inline app/controllers/questions_controller.rb]: Is the status implied?

File: app/models/question.rb (hunk 1 above)
Comment (andela-aadepoju) [inline app/models/question.rb]: :+1:

File: app/controllers/questions_controller.rb [hunk 3]
```diff
@@ -0,0 +1,49 @@
...
+  def show
+      render json: @question, status: 200
+  end
+
+  def create
+    question = Question.new(questions_params)
+    if question.save
+      render json: question, status: :created
+    else
+      render json:  question.errors, status: 400
```
Comment (andela-ooranagwa) [inline app/controllers/questions_controller.rb]: @andela-oojewale made some suggestions on how best to return `form errors`

File: app/controllers/questions_controller.rb [hunk 4]
```diff
@@ -0,0 +1,49 @@
...
+
+  def update
+    if @question.update(questions_params)
+      render json: @question, status: :ok
+    end
+  end
+
+  def destroy
+    if @question.try(:destroy)
+      render json: :no_content, status: 204
```
Comment (andela-ooranagwa) [inline app/controllers/questions_controller.rb]: Using `head 204` might be shorter

File: app/controllers/questions_controller.rb [hunk 5]
```diff
@@ -0,0 +1,49 @@
...
+    end
+  end
+
+  def top_questions
+    questions = Question.top
+    render json: questions
```
Comment (andela-ooranagwa) [inline app/controllers/questions_controller.rb]: Status missing

Comment (andela-aadepoju) [inline app/controllers/questions_controller.rb]: This has been fixed

File: docs/questions_endpoints.md [hunk 6]
```diff
@@ -196,10 +196,10 @@ Status: 404
   }
 ```
 
-## GET /questions/top_questions
+## GET /top_questions
 Request
 ```ruby
- GET  /questions/top_questions?limit=5,offset=1
+ GET  /top_questions?limit=5,offset=1
```
Comment (andela-ooranagwa) [inline docs/questions_endpoints.md]: > GET  /questions/top_questions?limit=5&offset=1

File: app/controllers/questions_controller.rb (hunk 3 above)
Comment (andela-aadepoju) [inline app/controllers/questions_controller.rb]: Where did he make the suggestions again?

File: app/controllers/questions_controller.rb [hunk 7]
```diff
@@ -0,0 +1,49 @@
...
+    else
+      render json:  question.errors, status: 400
+    end
+  end
+
+  def update
+    if @question.update(questions_params)
+      render json: @question, status: :ok
+    end
+  end
```
Comment (andela-ooranagwa) [inline app/controllers/questions_controller.rb]: Just wondering; what happens _if @question.update()_ is not successful? :thought_balloon:

**Comment by andela-ooranagwa** [comment]
The alternatives to most of the `conditionals` are not provided; I believe that providing them might be relevant is some cases. 
Aside that (and the other minor comments I shared above), nice work chief. Cooool.

File: app/controllers/questions_controller.rb [hunk 8]
```diff
@@ -0,0 +1,49 @@
...
+    question = Question.new(questions_params)
+    if question.save
+      render json: question, status: :created
+    else
+      render json:  question.errors, status: 400
+    end
+  end
+
+  def update
+    if @question.update(questions_params)
```
Comment (andela-oakinniranye) [inline app/controllers/questions_controller.rb]:
```ruby
if @question.update(question_params)
render json: @question, status: 200
else
render json: @question.errors, status: :unprocessable_entity
end
```
Handle the `conditions`, else the frontend may assume that the request was successful.

File: app/controllers/questions_controller.rb [hunk 9]
```diff
@@ -0,0 +1,49 @@
...
+  end
+
+  private
+
+  def set_question
+    @question = Question.find_by(id: params[:id])
+    resource_not_found && return unless @question
+  end
+
+  def questions_params
```
Comment (andela-oakinniranye) [inline app/controllers/questions_controller.rb]: this should be `question_params` not `questions_params`

File: app/controllers/questions_controller.rb [hunk 10]
```diff
@@ -0,0 +1,49 @@
...
+    params.permit(:id, :title, :content, :votes, :user_id)
```
Comment (andela-oakinniranye) [inline app/controllers/questions_controller.rb]: If we're authenticating a user via the `Authorization header`, we may not need to pass the `:user_id` params from the frontend anymore

File: app/controllers/questions_controller.rb [hunk 11]
```diff
@@ -0,0 +1,52 @@
...
+    else
+      invalid_request("field can't be blank")
+    end
+  end
+
+  def update
+    if @question.try(:update, questions_params)
+      render json: @question, status: 200
+    else
+       invalid_request("field can't be blank")
```
Comment (andela-oakinniranye) [inline app/controllers/questions_controller.rb]: `invalid_request("field can't be blank")` how about when the update couldn't be done due to other errors. `Field can't be blank is a bit generic`

**Comment by andela-oakinniranye** [comment]
:+1:

File: app/controllers/questions_controller.rb [hunk 12]
```diff
@@ -0,0 +1,57 @@
...
+  def top_questions
+    questions = Question.top
+    render json: questions, status: 200
+  end
+
+  private
+
+  def set_question
+    @question = Question.find_by(id: questions_params[:id])
+    resource_not_found && return unless @question
```
Comment (andela-echigbo) [inline app/controllers/questions_controller.rb]: Do we need the `return` statement? I think the `resource_not_found` method does  render of the response.

Comment (andela-aadepoju) [inline app/controllers/questions_controller.rb]: The return is necessary when if we need to add anything after that line in future. It doesn't harm to leave it there now too.

File: app/controllers/application_controller.rb [hunk 13]
```diff
@@ -1,17 +1,21 @@
...
   include ActionController::HttpAuthentication::Token::ControllerMethods
+  include ActionController::Serialization
+
 
   attr_reader :current_user
   before_action :authenticate_user
 
   def resource_not_found
-    render json: {errors: "The resource you tried to access was not found"}, status: 404
+    not_found = "The resource you tried to access was not found"
```
Comment (andela-ooranagwa) [inline app/controllers/application_controller.rb]: Nice! Keeping to style guide :+1:

**Pull request merged**